File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    factors
      |> Enum.reduce([], &(scores(limit, &1, &1, &2)))
      |> Enum.uniq
      |> Enum.sum
  end

  defp scores(limit, _factor, multiplier, acc) when limit <= multiplier, do: acc
  defp scores(limit,  factor, multiplier, acc) do
    scores(limit, factor, multiplier + factor, [multiplier | acc])
  end
end

File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.List (nub)

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum . nub $ scores
  where
    score limit factor multiplier acc
      | limit <= multiplier = acc
      | otherwise =
        score limit factor (multiplier + factor) (multiplier : acc)

    scores = foldl (\acc x -> score limit x x acc) [] factors
